Review adding AD as an external authentication source (#3149)
theforeman/foreman-documentation: guides/common/modules/proc_configuring-direct-ad-integration-with-gss-proxy.adoc

+
[options="nowrap", subs="+quotes,verbatim,attributes"]
----
[domain/_EXAMPLE_]
access_provider = ad
ad_gpo_access_control = enforcing
ad_gpo_map_service = +foreman
----
ifdef::satellite[]
+
For more information on GPOs, see link:{RHELDocsBaseURL}9/html/integrating_rhel_systems_directly_with_windows_active_directory/managing-direct-connections-to-ad_integrating-rhel-systems-directly-with-active-directory#how-sssd-interprets-gpo-access-control-rules_applying-group-policy-object-access-control-in-rhel[How SSSD interprets GPO access control rules] in _{RHEL}{nbsp}9 Integrating RHEL systems directly with Windows Active Directory_ or link:{RHELDocsBaseURL}8/html/integrating_rhel_systems_directly_with_windows_active_directory/managing-direct-connections-to-ad_integrating-rhel-systems-directly-with-active-directory#applying-group-policy-object-access-control-in-rhel_managing-direct-connections-to-ad[How SSSD interprets GPO access control rules] in _{RHEL}{nbsp}8 Integrating RHEL systems directly with Windows Active Directory_.
endif::[]
.. Restart the `sssd` service:
+
[options="nowrap", subs="+quotes,verbatim,attributes"]
----
# systemctl restart sssd
----

> . Enable IPA authentication in {Project}:

Reviewer: Looking at this, I doubt it's really correct. That feature is really designed for IPA, not for AD.

For example, it ensures the keytab is owned by Apache, which directly contradicts the statement that Apache shouldn't be able to read the keytab. However, it does set up some things in Apache that may be needed.

Big question: what is actually needed for the AD integration?

Reviewer 2: I believe in this case, AD is masked as IPA by SSSD. That's how I always imagined it to work.

Reviewer: Further elaborating on this: we explicitly tell Apache about its location: https://github.com/theforeman/puppet-foreman/blob/5f38479c870ef03a4a1565007be68b61e8950d53/templates/auth_gssapi.conf.erb#L6

Reviewer: It also inspired me to refactor a bit: https://github.com/theforeman/puppet-foreman/pull/1172

Reviewer: > I believe in this case, AD is masked as IPA by SSSD. That's how I always imagined it to work.

I'd classify it as a hack and fundamentally there are some differences here that IMHO make it incompatible. In particular, the keytab handling is different. At least when I read this documentation and the code for IPA.

Reviewer 2: Unaccessibility by Apache can still be achieved using permissions and selinux context on that file, right?

Reviewer 2: Does Apache only need to read it during installation or does it access it regularly? If the latter, we either need to change the code or the goal outlined in the docs can't be reached IMO.

Reviewer 2: Anyway, when I used path `/etc/httpd/config/http.conf` with `root.root`, `600` and `krb5_keytab_t`, it worked for me.

Reviewer 2: ```
# sealert -l 20295c58-da31-43c5-897c-63c5e83544e0
SELinux is preventing /usr/sbin/gssproxy from search access on the directory httpd.

*****  Plugin catchall (100. confidence) suggests   **************************

If you believe that gssproxy should be allowed search access on the httpd directory by default.
Then you should report this as a bug.
You can generate a local policy module to allow this access.
Do
allow this access for now by executing:
# ausearch -c 'gssproxy' --raw | audit2allow -M my-gssproxy
# semodule -X 300 -i my-gssproxy.pp

Additional Information:
Source Context                system_u:system_r:gssproxy_t:s0
Target Context                system_u:object_r:httpd_config_t:s0
Target Objects                httpd [ dir ]
Source                        gssproxy
Source Path                   /usr/sbin/gssproxy
Port                          <Unknown>
Host                         <FQDN>
Source RPM Packages           
Target RPM Packages           
SELinux Policy RPM            selinux-policy-targeted-3.14.3-139.el8_10.noarch
Local Policy RPM              selinux-policy-targeted-3.14.3-139.el8_10.noarch
Selinux Enabled               True
Policy Type                   targeted
Enforcing Mode                Enforcing
Host Name                     <FQDN>
Platform                      Linux
                              <FQDN>
                              4.18.0-553.8.1.el8_10.x86_64 #1 SMP Fri Jun 14
                              03:19:37 EDT 2024 x86_64 x86_64
Alert Count                   9
First Seen                    2024-07-25 05:10:52 EDT
Last Seen                     2024-07-25 09:12:22 EDT
Local ID                      20295c58-da31-43c5-897c-63c5e83544e0

Raw Audit Messages
type=AVC msg=audit(1721913142.873:1466): avc:  denied  { search } for  pid=27259 comm="gssproxy" name="httpd" dev="vda3" ino=159424064 scontext=system_u:system_r:gssproxy_t:s0 tcontext=system_u:object_r:httpd_config_t:s0 tclass=dir permissive=0

Hash: gssproxy,gssproxy_t,httpd_config_t,dir,search
```

Reviewer 2: And the autogenerated rule is:
```
module my-gssproxy 1.0;

require {
        type httpd_config_t;
        type gssproxy_t;
        class dir search;
}

#============= gssproxy_t ==============
allow gssproxy_t httpd_config_t:dir search;
```
That looks like gssproxy just isn't supposed to look into Apache's folder. It all looks hacky.
Surprisingly, it works.

Author: With the latest commit, I went with @lhellebr's "hacky but works" approach.

# id {apache-user}
----
+

> The Apache user must not have access to the keytab file.

Reviewer: The installer enforces that the Apache user does have access
```suggestion
```

cred_store = ccache:/var/lib/gssproxy/clients/krb5cc_%U
euid = __ID_of_Apache_User__
----
. Add a SELinux context rule to apply the appropriate security context to the `http.keytab` file:
+
[options="nowrap", subs="+quotes,verbatim,attributes"]
----
# semanage fcontext -a -s system_u -t krb5_keytab_t '/etc/httpd/conf/http\.keytab'
----
. Create a keytab entry:
+
[options="nowrap", subs="+quotes,verbatim,attributes"]
----
# KRB5_KTNAME=FILE:/etc/httpd/conf/http.keytab net ads keytab add HTTP -U administrator -d3 -s /etc/net-keytab.conf

> # chown root.root /etc/httpd/conf/http.keytab
> # chmod 600 /etc/httpd/conf/http.keytab

Reviewer: The installer enforces different permissions
```suggestion
```

cred_store = ccache:/var/lib/gssproxy/clients/krb5cc_%U
euid = __ID_of_Apache_User__
----
. Add a SELinux context rule to apply the appropriate security context to the `http.keytab` file:
+
[options="nowrap", subs="+quotes,verbatim,attributes"]
----

> # semanage fcontext -a -s system_u -t krb5_keytab_t '/etc/httpd/conf/http\.keytab'

Reviewer 2: There was a suggestion to move the keytab to /etc/gssproxy, why keep it in /etc/httpd/conf when the purpose of this setup is to " implement stricter privilege separation for the Apache server by removing access to the keytab file while preserving Kerberos authentication functionality"?

Reviewer 2: IMHO both locations (or any other) are good as long as correct permissions and SELinux context are set. I think with `/etc/gssproxy`, the semanage and restorecon commands may not be necessary.
Why not hold on to our own knowledge base though? https://access.redhat.com/articles/5854761

Author: Ok, let's stick with /etc/gssproxy to match the KBase article. I haven't dropped semanage because according to https://issues.redhat.com/browse/SAT-22855, it should be necessary. But I can drop it if you tell me to @lhellebr

I think I can resolve this thread for now.

> ----

Author: Requested in SAT-22855.

# chown root.{apache-user} /etc/httpd/conf/http.keytab
# chmod 640 /etc/httpd/conf/http.keytab
----

> . Configure SSSD to map Group Policy Objects (GPOs) from AD to the `foreman` PAM service:

Reviewer 2: Finally someone moved it from that confusing box that everyone ignored. Thanks!

.. Add the following lines to your `/etc/sssd/sssd.conf`:
+
[options="nowrap", subs="+quotes,verbatim,attributes"]
----
access_provider = ad
ad_gpo_access_control = enforcing
ad_gpo_map_service = +foreman
----
ifdef::satellite[]
+

> For more information on GPOs, see https://access.redhat.com/documentation/en-us/red_hat_enterprise_linux/8/html/integrating_rhel_systems_directly_with_windows_active_directory/managing-direct-connections-to-ad_integrating-rhel-systems-directly-with-active-directory#applying-group-policy-object-access-control-in-rhel_managing-direct-connections-to-ad[How SSSD interprets GPO access control rules] in _Integrating RHEL systems directly with Windows Active Directory_.

Reviewer 3: ```suggestion
For more information on GPOs, see {RHELDocsBaseURL}8/html/integrating_rhel_systems_directly_with_windows_active_directory/managing-direct-connections-to-ad_integrating-rhel-systems-directly-with-active-directory#applying-group-policy-object-access-control-in-rhel_managing-direct-connections-to-ad[How SSSD interprets GPO access control rules] in _Integrating RHEL systems directly with Windows Active Directory_.
```
Also, why remove the link to RHEL 9?

Author: Good catch, I didn't notice that happened. Re-adding the link.

endif::[]
.. Restart the `sssd` service:
+
[options="nowrap", subs="+quotes,verbatim,attributes"]
----
# systemctl restart sssd

> ----

Author: Moved from the now deleted WARNING box for better visibility. Also, added a step to restart SSSD.

> For more information on GPOs, see link:{RHELDocsBaseURL}9/html/integrating_rhel_systems_directly_with_windows_active_directory/managing-direct-connections-to-ad_integrating-rhel-systems-directly-with-active-directory#how-sssd-interprets-gpo-access-control-rules_applying-group-policy-object-access-control-in-rhel[How SSSD interprets GPO access control rules] in _{RHEL}{nbsp}9 Integrating RHEL systems directly with Windows Active Directory_ or link:https://{RHELDocsBaseURL}8/html/integrating_rhel_systems_directly_with_windows_active_directory/managing-direct-connections-to-ad_integrating-rhel-systems-directly-with-active-directory#applying-group-policy-object-access-control-in-rhel_managing-direct-connections-to-ad[How SSSD interprets GPO access control rules] in _{RHEL}{nbsp}8 Integrating RHEL systems directly with Windows Active Directory_.

Reviewer 3: ```suggestion
For more information on GPOs, see link:{RHELDocsBaseURL}9/html/integrating_rhel_systems_directly_with_windows_active_directory/managing-direct-connections-to-ad_integrating-rhel-systems-directly-with-active-directory#how-sssd-interprets-gpo-access-control-rules_applying-group-policy-object-access-control-in-rhel[How SSSD interprets GPO access control rules] in _{RHEL}{nbsp}9 Integrating RHEL systems directly with Windows Active Directory_ or link:{RHELDocsBaseURL}8/html/integrating_rhel_systems_directly_with_windows_active_directory/managing-direct-connections-to-ad_integrating-rhel-systems-directly-with-active-directory#applying-group-policy-object-access-control-in-rhel_managing-direct-connections-to-ad[How SSSD interprets GPO access control rules] in _{RHEL}{nbsp}8 Integrating RHEL systems directly with Windows Active Directory_.
```
:coffee: :)

----
# systemctl restart httpd
----

> . Add a *Service Principal Name (SPN)* in your AD server to enable Kerberos connections.

Reviewer 2: I've never had to do this since this command (from docs 5.3.3. Configuring Direct AD Integration with GSS-Proxy, step 6) adds Satellite as a client service - that is, adds it to AD and gets its keytab:
```
KRB5_KTNAME=FILE:/etc/httpd/conf/http.keytab net ads keytab add HTTP -U administrator -d3 -s /etc/net-keytab.conf
```

Reviewer 2: Of course use of `administrator` may be questionable...

> For more information, see link:https://learn.microsoft.com/en-us/sql/database-engine/configure-windows/register-a-service-principal-name-for-kerberos-connections?view=sql-server-ver16[Register a Service Principal Name for Kerberos connections].

Author: Requested in SAT-22855.
